Added "suggestion implemented" command.

DrAgenda opened the pull request:
A command that allows to designate a suggestion as implemented.

DrAgenda committed:
Added "suggestion implemented" command.

A command that allows to designate a suggestion as implemented.

Marco4413 commented on src/commands/Suggestion.js:
```diff
@@ -10,8 +10,9 @@ const { ReplyIfBlacklisted } = require("./utils/AccessListUtils.js");
+ * @param {Boolean} implemented
  */
-const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, approve) => {
+const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, approve, implemented) => {
```
Using two booleans to do that will create bugs. As of right now, the bot would always send two messages if `implemented` is set to `true`. I would recommend using an enum.

Marco4413 commented on src/localization/en-us/commands.json:
```diff
@@ -346,6 +346,9 @@
     "suggestionRejectedColor": "0xcc0000",
     "suggestionRejectedTitle": "Suggestion Rejected.",
     "suggestionRejectedDescription": "A suggestion sent by {user-mention} was rejected.",
+    "suggestionImplementedColor": "1030B0",
```
Not sure if Discord accepts hex without the `0x` prefix. Even if they work, I would suggest changing them to fit the convention used by the other ones.

Marco4413 replied:
It's not actually whether or not Discord supports it, the color is given to `Number.parseInt` and that function needs `0x` to know that the number is hex.

Marco4413 commented on src/localization/en-us/commands.json:
```diff
@@ -346,6 +346,9 @@
     "suggestionRejectedDescription": "A suggestion sent by {user-mention} was rejected.",
+    "suggestionImplementedColor": "1030B0",
+    "suggestionImplementedTitle": "Suggestion Designated As Implemented.",
+    "suggestionImplementedDescription": "A suggestion sent by {user-mention} was designated as implemented.",
```
I'd prefer the text "Suggestion Already Implemented." as the title and "A suggestion sent by {user-mention} was already implemented." for the description.

Marco4413 commented on src/localization/it-it/commands.json:
```diff
@@ -346,6 +346,9 @@
     "suggestionRejectedDescription": "Un suggerimento inviato da {user-mention} è stato rifiutato.",
+    "suggestionImplementedColor": "1030B0",
+    "suggestionImplementedTitle": "Suggerimento Designato Come Implementato.",
+    "suggestionImplementedDescription": "Un suggerimento inviato da {user-mention} è stato designato come implementato.",
```
Same issues as the English locale.

Marco4413 reviewed (changes requested):
There are some issues with this PR. I have put some comments inside the files you have changed explaining the issues that are present.

A brief list of the problems I have found:
- [x] Duplicate messages sent if `implemented` is `true` on `_ProcessSuggestion` call.
- [x] Follow the convention that is already in use by colors inside locale files.
- [x] Improve title and description for response embeds.

DrAgenda committed:
Fixed "suggestion implemented" command.

Fixed "suggestion implemented" command:
- Improved embed descriptions;
- Fixed "_ProcessSuggestion";
- Changed colors in locale file.

Marco4413 commented on src/commands/Suggestion.js:
```diff
@@ -57,30 +64,28 @@ const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, appr
-    }
 
-    if (approve) {
+    if (SuggestionStatus.Approved) {
```
No switch statement?

Marco4413 reviewed (changes requested):
Eslint action is failing and jsdoc comment for `_ProcessSuggestion` is wrong.

Remaining Issues:
- [ ] Eslint action is failing.
- [ ] jsdoc comment for `_ProcessSuggestion` is incorrect.
- [x] A switch statement can be used instead of multiple if statements.
- [x] Enum values are strings (checking for string equality is slower than numbers).

DrAgenda committed:
Fix number 3 of "suggestion implemented" command.

Fixed "_ProcessSuggestion" (again).

Marco4413 commented on src/commands/Suggestion.js:
```diff
@@ -88,6 +86,33 @@ const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, Sugg
     }
 
+    switch (SuggestionStatus) {
+        case 1:
```
Indentation doesn't respect the standard of the project.
See:
https://github.com/Marco4413/JackBot/blob/a7739bf723d9a942442cff722c862a6ac549057e/src/commands/Sticky.js#L40-L52

Marco4413 commented on src/commands/Suggestion.js:
```diff
@@ -9,14 +9,12 @@ const { ReplyIfBlacklisted } = require("./utils/AccessListUtils.js");
+    Implemented: 3
 }
 
 const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, SuggestionStatus) => {
```
Argument `SuggestionStatus` shadows the enum definition.

Marco4413 commented on src/commands/Suggestion.js:
```diff
@@ -88,6 +86,33 @@ const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, Sugg
+                { "user-mention": Utils.MentionUser(suggestionRow.authorId) }
+            ));
+            break;
+        case 2:
```
Using hard-coded numbers defeats the purpose of an enum.

Marco4413 commented on src/commands/Suggestion.js:
```diff
@@ -88,6 +86,33 @@ const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, Sugg
+                { "user-mention": Utils.MentionUser(suggestionRow.authorId) }
+            ));
+            break;
+    }
```
Switch has no `default` branch. I would suggest having `SuggestionStatus.Approved` be the default behaviour, or log the invalid state.

Marco4413 commented:
Your switch statement has some issues:
- [ ] Doesn't follow the project's convention for indentation.
- [ ] Uses hard-coded number instead of the values within the enum.
- [ ] Has no `default` branch.

The function `_ProcessSuggestion` has the following issues:
- [ ] Argument `SuggestionStatus` shadows the enum definition and, therefore, cannot be used inside the switch statement.

DrAgenda committed:
Fixed "suggestion implemented" command.

Fixed "_ProcessSuggestion" call:
- Switch statement for status;
- Enum for status.

Marco4413 commented on src/commands/Suggestion.js:
```diff
@@ -63,54 +64,33 @@ const _ProcessSuggestion = async (msg, guild, locale, suggestionId, reason, Sugg
             ));
             break;
-        case 3:
+        case status.Implemented:
```
`status.Implemented` is undefined, same for the other cases. This won't work.

Marco4413 commented:
This feature was added with the latest commit (5b469b385a8c0434a4493c707829b6eb411d30b9).
